spirv_hlsl.cpp

```diff
@@ -1508,7 +1512,45 @@ void CompilerHLSL::emit_buffer_block(const SPIRVariable &var)
 
 void CompilerHLSL::emit_push_constant_block(const SPIRVariable &var)
 {
-	emit_buffer_block(var);
+	std::cout << options.root_constants_layout.size() << std::endl;
+	if (options.root_constants_layout.empty())
+	{
+		emit_buffer_block(var);
+		return;
+	}
+	else
+	{
+		for(auto const& layout : options.root_constants_layout)
+		{
+			auto &type = get<SPIRType>(var.basetype);
+
+			flattened_structs.insert(var.self);
+			type.member_name_cache.clear();
+			add_resource_name(var.self);
+			auto &memb = meta[type.self].members;
+
+			statement("cbuffer ", to_name(var.self), ": register(b", layout.start / 4, ", space0)");
+			begin_scope();
+			for (auto offset = layout.start; offset < layout.end; offset += 4)
+			{
+				uint32_t i = 0;
+				for (auto &member : type.member_types)
+				{
+					if (memb[i].offset == offset)
+					{
+						add_member_name(type, i);
+						auto backup_name = get_member_name(type.self, i);
+						auto member_name = to_member_name(type, i);
+						set_member_name(type.self, i, sanitize_underscores(join(to_name(type.self), "_", member_name)));
+						emit_struct_member(type, member, i, "", layout.start);
```

Comment:
How does this deal with buffer blocks with "odd" layouts where manual packoffset() must be used?

Reply:
Could you give an example for these odd layouts? It might not apply for push constants but I'm not sure atm.

Reply:
A classic example here is:
```
push_constant {
   float foo; // Offset 0
   vec3 bar; // Offset 16 in GLSL, 4 in HLSL
}
```
The main implementation of buffer blocks validates the buffer blocks for offsets and uses manual layout(offset) or packoffset() as necessary.

There are also cases where packoffset is impossible, for example substructs within a cbuffer. If there is a mismatch between the SPIR-V and equivalent HLSL packing for that substruct, an exception is thrown.

Reply:
Thanks for bringing this point up! Played a bit around with it. I works depending on the specified root constant ranges. There are certain cases where I see no way to work around it, but I need to double-check it.
For example: 
```
push_constant {
   float foo; // VERTEX shader only, other fields VERTEX and FRAGMENT shader

   float foo1;
   float foo2;
   float foo3;
   vec3 bar; // Misalignment here
}
```
Splitting after the first float will result in misaligment of `bar` in the fragment shader.
Splitting at offset 16 in the example you mentioned would be fine on the other hand.